planetiler-custommap/src/main/java/com/onthegomap/planetiler/custommap/ConfiguredFeature.java

```diff
@@ -0,0 +1,317 @@
+      return null;
+    }
+
+    Function<SourceFeature, Byte> staticZooms =
+      sf -> (byte) Math.max(minZoom,
+        minZoomFromTilePercent(sf, minTilePercent));
+
+    if (minZoomByValue.isEmpty()) {
+      return (sf, key) -> staticZooms.apply(sf);
+    }
+
+    //Attribute value-specific zooms override static zooms
+    return (sourceFeature, key) -> minZoomByValue.getOrDefault(key, staticZooms.apply(sourceFeature));
+  }
+
+  private static byte minZoomFromTilePercent(SourceFeature sf, Double minTilePercent) {
+    if (minTilePercent == null) {
+      return 0;
+    }
+    try {
+      return (byte) (Math.log(minTilePercent / sf.area()) / LOG4);
+    } catch (GeometryException e) {
+      return 14;
+    }
+  }
+
+  /**
+   * Generates a function which produces a fully-configured attribute for a feature.
+   * 
+   * @param attribute - configuration for this attribute
+   * @return processing logic
+   */
+  private BiConsumer<SourceFeature, Feature> attributeProcessor(AttributeDefinition attribute) {
+    var tagKey = attribute.key();
+
+    var attributeMinZoom = attribute.minZoom();
+    attributeMinZoom = attributeMinZoom == null ? 0 : attributeMinZoom.byteValue();
+
+    var minZoomByValue = attribute.minZoomByValue();
+    minZoomByValue = minZoomByValue == null ? Map.of() : minZoomByValue;
+
+    //Workaround because numeric keys are mapped as String
+    minZoomByValue = tagValueProducer.remapKeysByType(tagKey, minZoomByValue);
+
+    var attributeValueProducer = attributeValueProducer(attribute);
+
+    var attrIncludeWhen = attribute.includeWhen();
+    var attrExcludeWhen = attribute.excludeWhen();
+
+    var attributeTest =
+      Expression.and(
+        attrIncludeWhen == null ? Expression.TRUE : matcher(attrIncludeWhen, tagValueProducer),
+        attrExcludeWhen == null ? Expression.TRUE : Expression.not(matcher(attrExcludeWhen, tagValueProducer))
+      );
+
+    var minTileCoverage = attrIncludeWhen == null ? null : attribute.minTileCoverSize();
+
+    BiFunction<SourceFeature, Object, Byte> attributeZoomProducer =
+      attributeZoomThreshold(minTileCoverage, attributeMinZoom, minZoomByValue);
+
+    if (attributeZoomProducer != null) {
+      return (sf, f) -> {
+        if (attributeTest.evaluate(sf)) {
+          Object value = attributeValueProducer.apply(sf);
+          f.setAttrWithMinzoom(tagKey, value, attributeZoomProducer.apply(sf, value));
+        }
+      };
+    }
+
+    return (sf, f) -> {
+      if (attributeTest.evaluate(sf)) {
+        f.setAttr(tagKey, attributeValueProducer.apply(sf));
+      }
+    };
+  }
+
+  /**
+   * Generates a factory method which creates a {@link Feature} from a {@link FeatureCollector} of the appropriate
+   * geometry type.
+   * 
+   * @param layerName - name of the layer
+   * @param type      - type of geometry
+   * @return geometry factory method
+   */
+  private Function<FeatureCollector, Feature> geometryMapFeature(String layerName, GeometryType type) {
```

Comment:
Can be static - also we might want to pull these into the constructors for each `GeometryType` 🤔   ... feel free to ignore though, that could be a future refactor as it doesn't impact internal or external API.